src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala

```diff
@@ -29,154 +29,121 @@ class TaskReplaceActor(
     val eventBus: EventStream,
     val readinessCheckExecutor: ReadinessCheckExecutor,
     val runSpec: RunSpec,
-    promise: Promise[Unit]) extends Actor with Stash with ReadinessBehavior with StrictLogging {
+    promise: Promise[Unit]) extends Actor with StrictLogging {
   import TaskReplaceActor._
 
   def deploymentId = status.plan.id
+  def pathId = runSpec.id
+
+  // instance to kill sorted by decreasing order of priority
+  // we always prefer to kill unhealthy tasks first
+  private[this] val toKillOrdered = instanceTracker.specInstancesSync(runSpec.id, readAfterWrite = true)
+    .partition(_.runSpecVersion == runSpec.version)._2.sortWith((i1, i2) => {
+      (i1.consideredHealthy, i2.consideredHealthy) match {
+        case (_, false) => false
+        case _ => true
+      }
+    })
 
-  // All existing instances of this app independent of the goal.
-  //
-  // Killed resident tasks are not expunged from the instances list. Ignore
-  // them. LaunchQueue takes care of launching instances against reservations
-  // first
-  val currentInstances = instanceTracker.specInstancesSync(runSpec.id, readAfterWrite = true)
-
-  // In case previous master was abdicated while the deployment was still running we might have
-  // already started some new tasks.
-  // All already started and active tasks are filtered while the rest is considered
-  private[this] val (instancesAlreadyStarted, oldInstances) = {
-    currentInstances.partition(_.runSpecVersion == runSpec.version)
-  }
+  // All instances to kill queued up
+  private[this] val toKill: mutable.Queue[Instance.Id] = toKillOrdered.map(_.instanceId).to[mutable.Queue]
```

Comment:
I know it comes from my code but it seems we are ordering once at the start of the actor and never consider that HC could have changed.
It is still an improvement compared to before but might be a problem if this actor lifetime is long enough (which is the case I think if the deployment last long)

Reply:
Yeah I thought about this case but in a first version I did not want to change to much. If I have some time at the end I will also compute this in the step() method.